File: Practice/Data Structures/Segment Tree.cpp
#include<bits/stdc++.h>

using namespace std;

class SegmentTree {
    vector<int> arr;
    int MAX_SIZE;
    int size;
    int single(int x) {
        return x;
    }
    int merge(int a, int b) {
        return a + b;
    }
    void buildSegmentTree(vector<int> &a, int root, int x, int y) {
        if(x < y) {
            if(y - x == 1) {
                if(x < size) {
                    arr[root] = single(a[x]);
                }
            } else {
                int m = (x + y) / 2;
                buildSegmentTree(a, root * 2 + 1, x, m);
                buildSegmentTree(a, root * 2 + 2, m, y);
                arr[root] = merge(arr[root*2+1], arr[root*2+2]);
            }
        }
    }
    void update(int root, int x, int y, int idx, int v) {
        if(x < y) {
            if(y - x == 1) {
                if(x == idx) {
                    arr[root] = single(v);
                }
            } else if(x <= idx && y > idx) {
                int m = (x + y) / 2;
                if(idx < m) {
                    update(root * 2 + 1, x, m, idx, v);
                } else {
                    update(root * 2 + 2, m, y, idx, v);
                }
                arr[root] = merge(arr[root*2+1], arr[root*2+2]);
            }
        }
    }
    int query(int root, int x, int y, int l, int r) {
        if(x < y) {
            if(y <= l || r <= x || x >= y) {
                return 0;
            } else if(y > l && x < r) {
                return arr[root];
            } else {
                int m = (x + y) / 2;
                return merge(query(root*2+1, x, m, l, r), query(root*2+2, m, y, l, r));
            }
        }
    }
public:
    SegmentTree(int n) { // Initialize with a n-length zer filled array
        MAX_SIZE = 1;
        size = n;
        while(MAX_SIZE < size) {
            MAX_SIZE <<= 1;
        }
        arr.assign(MAX_SIZE << 1, 0);
    }
    SegmentTree(vector<int> &a) : SegmentTree(a.size()) {
        buildSegmentTree(a, 0, 0, MAX_SIZE);
    }
    void update(int idx, int v) {
        update(0, 0, MAX_SIZE, idx, v);
    }
    int query(int l, int r) {
        return query(0, 0, MAX_SIZE, l, r);
    }
    void print() {
        for(int i: arr) {
            cout << i << ' ';
        }
        cout << '\n';
    }
};

int main() {
    vector<int> a({8, 7, 6, 5, 4, 3, 2, 1});
    SegmentTree st(a);
    st.print();
    for(int i = 0; i < 8; ++i) {
        st.update(i, i+1);
        st.print();
    }
    return 0;
}

File: Practice/Golang/Data Structures/segment tree.go
package main

import (
	"fmt"
	"math"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

var segmentTree []int
var sizeSegmentTree int

func initSegmentTree(arr []int, n int) {
	sizeSegmentTree = 1
	for sizeSegmentTree < n {
		sizeSegmentTree = sizeSegmentTree * 2
	}
	segmentTree = make([]int, sizeSegmentTree*2-1)
	for i := range segmentTree {
		segmentTree[i] = math.MaxInt32
	}
	buildSegmentTree(arr, 0, 0, sizeSegmentTree-1)
}

func buildSegmentTree(arr []int, root, l, r int) {
	if l == r {
		if r < len(arr) {
			segmentTree[root] = arr[l]
		}
	} else if l < r {
		m := (l + r) / 2
		buildSegmentTree(arr, root*2+1, l, m)
		buildSegmentTree(arr, root*2+2, m+1, r)
		segmentTree[root] = min(segmentTree[root*2+1], segmentTree[root*2+2])
	}
}

func updateSegmentTree(root, l, r, idx, v int) {
	if l <= r && l <= idx && r >= idx {
		if l == r {
			if l == idx {
				segmentTree[root] = v
			}
		} else {
			m := (l + r) / 2
			if idx > m {
				updateSegmentTree(root*2+2, m+1, r, idx, v)
			} else {
				updateSegmentTree(root*2+1, l, m, idx, v)
			}
			segmentTree[root] = min(segmentTree[root*2+1], segmentTree[root*2+2])
		}
	}
}

func querySegmentTree(root, l, r, x, y int) int {
	if r < l || y < x || r < x || y < l {
		return math.MaxInt32
	}
	if x <= l && r <= y {
		return segmentTree[root]
	}
	m := (l + r) / 2
	return min(querySegmentTree(root*2+1, l, m, x, y), querySegmentTree(root*2+2, m+1, r, x, y))
}

func main() {
	var n int
	fmt.Scanf("%d", &n)
	var arr []int = make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scanf("%d", &arr[i])
	}
	initSegmentTree(arr, n)
	fmt.Println(segmentTree)
	var m int
	fmt.Scanf("%d", &m)
	for i := 0; i < m; i++ {
		var q int
		fmt.Scanf("%d", &q)
		if q == 1 {
			var idx, v int
			fmt.Scanf("%d %d", &idx, &v)
			updateSegmentTree(0, 0, sizeSegmentTree-1, idx, v)
			fmt.Println(segmentTree)
		} else {
			var l, r int
			fmt.Scanf("%d %d", &l, &r)
			fmt.Println(querySegmentTree(0, 0, sizeSegmentTree - 1, l, r))
		}

	}

}
